libvast/src/segment_store.cpp

```diff
@@ -177,27 +170,135 @@ std::unique_ptr<store::lookup> segment_store::extract(const ids& xs) const {
+        VAST_WARNING(this, "got no slices after lookup for segment", segment_id,
+                     "=> erases entire segment!");
+        erased_events += drop(sref);
+        return;
+      }
+    } else {
+      VAST_WARNING(this, "was unable to get table slice for segment",
+                   segment_id, "=> erases entire segment!");
+      erased_events += drop(sref);
+      return;
+    }
+    VAST_ASSERT(slices.size() > 0);
+    auto keep_mask = ~xs;
+    std::vector<table_slice_ptr> new_slices;
+    for (auto& slice : slices) {
+      // Expand keep_mask on-the-fly if needed.
+      auto max_id = slice->offset() + slice->rows();
+      if (keep_mask.size() < max_id)
+        keep_mask.append_bits(true, max_id - keep_mask.size());
+      size_t new_slices_size_before = new_slices.size();
+      select(new_slices, slice, keep_mask);
+      size_t remaining_rows = 0;
+      for (size_t i = new_slices_size_before; i < new_slices.size(); ++i)
+        remaining_rows += new_slices[i]->rows();
+      erased_events += slice->rows() - remaining_rows;
+    }
+    if (new_slices.empty()) {
+      VAST_WARNING(this, "was unable to generate any new slice for segment",
+                   segment_id, "=> erases entire segment!");
+      erased_events += drop(sref);
+      return;
+    }
+    VAST_DEBUG(this, "shrinks segment", segment_id, "from", slices.size(), "to",
+               new_slices.size(), "slices");
+    // Remove stale state.
+    segments_.erase_value(segment_id);
+    // Create a new segment from the remaining slices.
+    segment_builder tmp_builder;
+    segment_builder* builder = &tmp_builder;
+    if constexpr (std::is_same_v<decltype(sref), segment_builder&>) {
+      // If `update` got called with a builder then we simply use that by
+      // resetting it and filling it with new content. Otherwise, we fill
+      // `tmp_builder` instead and replace the the segment `sref` in the next
+      // `if constexpr` block.
+      sref.reset();
+      builder = &sref;
+    }
+    for (auto& slice : new_slices) {
+      if (auto err = builder->add(slice)) {
+        VAST_ERROR(this, "failed to add slice to builder:" << err);
+      } else if (!segments_.inject(slice->offset(),
+                                   slice->offset() + slice->rows(),
+                                   builder->id()))
+        VAST_ERROR(this, "failed to update range_map");
+    }
+    // Flush the new segment and remove the previous segment.
+    if constexpr (std::is_same_v<decltype(sref), segment&>) {
+      auto new_segment = builder->finish();
+      auto filename = segment_path() / to_string(new_segment->id());
+      if (auto err = save(nullptr, filename, new_segment))
+        VAST_ERROR(this, "failed to persist the new segment");
+      auto stale_filename = segment_path() / to_string(segment_id);
+      // Schedule deletion of the segment file when releasing the chunk.
+      sref.chunk()->add_deletion_step([=] { rm(stale_filename); });
```

Comment:
I'm a bit hesitant to drill into the API of `chunk` for what we're trying to do here. Admittedly, I don't see another good way of solving it. @tobim, do you have an idea?

Reply:
This code needs to run when running the destructor of the chunk. So I think `add_deletion_step` is the most natural and direct way to express this.